archiver/worker.go

```diff
@@ -0,0 +1,264 @@
+			jobs,
+			jobsdb.Aborted.State,
+			errJson(fmt.Errorf("%s archival disabled for workspace %s", w.archiveFrom, workspaceID)),
+		); err != nil {
+			if w.lifecycle.ctx.Err() != nil {
+				return false
+			}
+			log.Errorw("failed to mark archive disabled jobs' status", "error", err)
+			panic(err)
+		}
+		if !limitReached {
+			return true
+		}
+		goto start
+	}
+
+	location, err := w.uploadJobs(w.lifecycle.ctx, jobs)
+	if err != nil {
+		log.Errorw("failed to upload jobs", "error", err)
+		return false
+	}
+	w.lastUploadTime = time.Now()
+
+	if err := w.markStatus(
+		jobs,
+		jobsdb.Succeeded.State,
+		locationJson(location),
+	); err != nil {
+		if w.lifecycle.ctx.Err() != nil {
+			return false
+		}
+		log.Errorw("failed to mark successful upload status", "error", err)
+		panic(err)
+	}
+	if !limitReached {
+		return true
+	}
+	goto start
+}
+
+func (w *worker) SleepDurations() (min, max time.Duration) {
+	if w.lastUploadTime.IsZero() {
+		return w.config.minSleep, w.config.uploadFrequency
+	}
+	return w.config.minSleep, time.Until(w.lastUploadTime.Add(w.config.uploadFrequency))
+}
+
+func (w *worker) Stop() {
+	w.lifecycle.cancel()
+}
+
+func (w *worker) uploadJobs(ctx context.Context, jobs []*jobsdb.JobT) (string, error) {
+	defer w.uploadLimiter.Begin(w.sourceID)()
+	firstJobCreatedAt := jobs[0].CreatedAt.UTC()
+	lastJobCreatedAt := jobs[len(jobs)-1].CreatedAt.UTC()
+	workspaceID := jobs[0].WorkspaceId
+	log := w.log.With("workspaceID", workspaceID)
+
+	gzWriter := fileuploader.NewGzMultiFileWriter()
+	path := fmt.Sprintf(
+		"%v%v.json.gz",
+		lo.Must(misc.CreateTMPDIR())+"/rudder-backups/"+w.sourceID+"/",
+		fmt.Sprintf("%v_%v_%v", firstJobCreatedAt.Unix(), lastJobCreatedAt.Unix(), workspaceID),
+	)
+
+	for _, job := range jobs {
+		j, err := marshalJob(job)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal job: %w", err)
+		}
+		if _, err := gzWriter.Write(path, append(j, '\n')); err != nil {
+			return "", fmt.Errorf("failed to write to file: %w", err)
+		}
+	}
+	if err := gzWriter.Close(); err != nil {
+		return "", fmt.Errorf("failed to close file: %w", err)
+	}
+	defer func() { _ = os.Remove(path) }()
+
+	fileUploader, err := w.storageProvider.GetFileManager(workspaceID)
+	if err != nil {
+		log.Errorw("Skipping storing errors since no file manager is found", "error", err)
+		return "", fmt.Errorf("failed to get file manager: %w", err)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to open file %s: %w", path, err)
+	}
+	defer func() { _ = file.Close() }()
+	prefixes := []string{
+		w.sourceID,
+		w.archiveFrom,
+		firstJobCreatedAt.Format("2006-01-02"),
+		fmt.Sprintf("%d", firstJobCreatedAt.Hour()),
+		w.config.instanceID,
+	}
+	uploadOutput, err := fileUploader.Upload(ctx, file, prefixes...)
+	if err != nil {
+		log.Errorw("failed to upload file to object storage", "error", err)
+		return "", fmt.Errorf("failed to upload file to object storage - %w", err)
```

Comment:
both logging and returning error

also:
```suggestion
		log.Errorw("failed to upload file to object storage", "error", err)
		return "", fmt.Errorf("upload file to object storage: %w", err)
```